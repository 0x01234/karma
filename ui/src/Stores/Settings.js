import { localStored } from "mobx-stored";

import { action } from "mobx";

const defaultSavedFilters = {
  filters: [],
  present: false
};

class Settings {
  savedFilters = localStored("savedFilters", defaultSavedFilters, {
    delay: 100
  });

  saveFilters = action(newFilters => {
    this.savedFilters.filters = newFilters;
    this.savedFilters.present = true;
  });

  clearSavedFilters = action(() => {
    this.savedFilters.filters = [];
    this.savedFilters.present = false;
  });
}

export { Settings };
