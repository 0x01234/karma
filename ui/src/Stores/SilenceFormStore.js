import { observable, action, computed } from "mobx";

import uniqueId from "lodash.uniqueid";

import moment from "moment";

const NewEmptyMatcher = id => {
  return {
    id: uniqueId(),
    name: "",
    values: [],
    suggestions: {
      names: [],
      values: []
    },
    isRegex: false
  };
};

const ValueToObject = value => ({ label: value, value: value });

class SilenceFormStore {
  // this is used to store modal visibility toggle
  toggle = observable(
    {
      visible: false,
      toggle() {
        this.visible = !this.visible;
      },
      hide() {
        this.visible = false;
      },
      show() {
        this.visible = true;
      }
    },
    { toggle: action.bound, hide: action.bound, show: action.bound }
  );

  // form data is stored here, it's global (rather than attached to the form)
  // so it can be manipulated from other parts of the code
  // example: when user clicks a silence button on alert we should populate
  // this form from that alert so user can easily silence that alert
  data = observable(
    {
      inProgress: false,
      alertmanagers: [],
      matchers: [],
      startsAt: moment(),
      endsAt: moment().add(1, "hour"),
      comment: "",
      author: "",

      resetProgress() {
        this.inProgress = false;
      },

      // append a new empty matcher to the list
      addEmptyMatcher() {
        let m = NewEmptyMatcher();
        this.matchers.push(m);
      },

      deleteMatcher(id) {
        // only delete matchers if we have more than 1
        if (this.matchers.length > 1) {
          this.matchers = this.matchers.filter(m => m.id !== id);
        }
      },

      fillMatchersFromGroup(group) {
        let matchers = [];

        // add matchers for all shared labels in this group
        for (const [key, value] of Object.entries(
          Object.assign({}, group.labels, group.shared.labels)
        )) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [ValueToObject(value)],
            suggestions: {
              names: [],
              values: []
            },
            isRegex: false
          });
        }

        // add matchers for all unique labels in this group
        let labels = {};
        for (const alert of group.alerts) {
          for (const [key, value] of Object.entries(alert.labels)) {
            if (!labels[key]) {
              labels[key] = new Set();
            }
            labels[key].add(value);
          }
        }
        for (const [key, values] of Object.entries(labels)) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [...values].sort().map(value => ValueToObject(value)),
            suggestions: {
              names: [],
              values: []
            },
            isRegex: values.size > 1
          });
        }

        this.matchers = matchers;
      },

      verifyStarEnd() {
        if (this.endsAt.isSameOrBefore(this.startsAt)) {
          this.endsAt = moment(this.startsAt).add(1, "minutes");
        }
      },
      incStart(minutes) {
        this.startsAt = moment(this.startsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decStart(minutes) {
        this.startsAt = moment(this.startsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incEnd(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decEnd(minutes) {
        this.endsAt = moment(this.endsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incDuration(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
      },
      decDuration(minutes) {
        const newEndsAt = moment(this.endsAt).subtract(minutes, "minutes");
        if (newEndsAt.isAfter(this.startsAt)) {
          this.endsAt = newEndsAt;
        }
      },

      get toAlertmanagerPayload() {
        const payload = {
          matchers: this.matchers.map(m => ({
            name: m.name,
            value:
              m.values.length > 1
                ? `(${m.values.map(v => v.value).join("|")})`
                : m.values.length === 1
                  ? m.values[0].value
                  : "",
            isRegex: m.isRegex
          })),
          startsAt: this.startsAt
            .second(0)
            .millisecond(0)
            .toISOString(),
          endsAt: this.endsAt
            .second(59)
            .millisecond(0)
            .toISOString(),
          createdBy: this.author,
          comment: this.comment
        };
        return payload;
      },
      get toDuration() {
        const data = {
          days: this.endsAt.diff(this.startsAt, "days"),
          hours: this.endsAt.diff(this.startsAt, "hours") % 24,
          minutes: this.endsAt.diff(this.startsAt, "minutes") % 60
        };
        return data;
      }
    },
    {
      resetProgress: action.bound,
      addEmptyMatcher: action.bound,
      deleteMatcher: action.bound,
      fillMatchersFromGroup: action.bound,
      verifyStarEnd: action.bound,
      incStart: action.bound,
      decStart: action.bound,
      incEnd: action.bound,
      decEnd: action.bound,
      incDuration: action.bound,
      decDuration: action.bound,
      toAlertmanagerPayload: computed,
      toDuration: computed
    },
    { name: "Silence form store" }
  );
}

export { SilenceFormStore };
