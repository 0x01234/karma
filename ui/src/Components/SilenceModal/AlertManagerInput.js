import React from "react";
import PropTypes from "prop-types";

import { action } from "mobx";
import { observer } from "mobx-react";

import ReactSelect from "react-select";

import { AlertStore } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { MultiSelect, ReactSelectStyles } from "Components/MultiSelect";
import { ValidationError } from "Components/MultiSelect/ValidationError";

const AlertmanagerInstancesToOptions = instances =>
  instances.map(i => ({
    label: i.name,
    value: i.publicURI
  }));

const AlertManagerInput = observer(
  class AlertManagerInput extends MultiSelect {
    static propTypes = {
      alertStore: PropTypes.instanceOf(AlertStore).isRequired,
      silenceFormStore: PropTypes.instanceOf(SilenceFormStore).isRequired
    };

    constructor(props) {
      super(props);

      const { alertStore, silenceFormStore } = props;

      if (silenceFormStore.data.alertmanagers.length === 0) {
        silenceFormStore.data.alertmanagers = AlertmanagerInstancesToOptions(
          alertStore.data.upstreams.instances
        );
      }
    }

    onChange = action((newValue, actionMeta) => {
      const { silenceFormStore } = this.props;

      silenceFormStore.data.alertmanagers = newValue;
    });

    componentDidUpdate() {
      const { alertStore, silenceFormStore } = this.props;

      // get the list of last known alertmanagers
      const currentAlertmanagers = AlertmanagerInstancesToOptions(
        alertStore.data.upstreams.instances
      );

      // now iterate what's set as silence form values and reset it if any
      // mismatch is detected (uri changed for example)
      for (const silenceAM of silenceFormStore.data.alertmanagers) {
        for (const currentAM of currentAlertmanagers) {
          if (
            silenceAM.label === currentAM.label &&
            silenceAM.value !== currentAM.value
          ) {
            silenceFormStore.data.alertmanagers = AlertmanagerInstancesToOptions(
              alertStore.data.upstreams.instances
            );
            return;
          }
        }
      }
    }

    render() {
      const { alertStore, silenceFormStore } = this.props;

      return (
        <ReactSelect
          styles={ReactSelectStyles}
          instanceId="silence-input-alertmanagers"
          defaultValue={silenceFormStore.data.alertmanagers}
          options={AlertmanagerInstancesToOptions(
            alertStore.data.upstreams.instances
          )}
          placeholder={
            silenceFormStore.data.wasValidated ? (
              <ValidationError />
            ) : (
              "Alertmanager"
            )
          }
          isMulti
          onChange={this.onChange}
        />
      );
    }
  }
);

export { AlertManagerInput };
