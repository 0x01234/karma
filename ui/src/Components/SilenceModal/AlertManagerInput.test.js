import React from "react";

import { shallow, mount } from "enzyme";

import { AlertStore } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { AlertManagerInput } from "./AlertManagerInput";

let alertStore;
let silenceFormStore;

const AlertmanagerOption = index => ({
  label: `am${index}`,
  value: `http://am${index}.example.com`
});

beforeEach(() => {
  alertStore = new AlertStore([]);
  alertStore.data.upstreams.instances = [
    { name: "am1", uri: "http://am1.example.com", error: "" },
    { name: "am2", uri: "http://am2.example.com", error: "" },
    { name: "am3", uri: "http://am3.example.com", error: "" }
  ];
  silenceFormStore = new SilenceFormStore();
});

const ShallowAlertManagerInput = () => {
  return shallow(
    <AlertManagerInput
      alertStore={alertStore}
      silenceFormStore={silenceFormStore}
    />
  );
};

const MountedAlertManagerInput = () => {
  return mount(
    <AlertManagerInput
      alertStore={alertStore}
      silenceFormStore={silenceFormStore}
    />
  );
};

const ValidateSuggestions = () => {
  const tree = MountedAlertManagerInput();
  // clear all selected instances, they are selected by default
  const clear = tree.find("ClearIndicator");
  // https://github.com/JedWatson/react-select/blob/c22d296d50917e210836fb011ae3e565895e6440/src/__tests__/Select.test.js#L1873
  clear.simulate("mousedown", { button: 0 });
  // click on the react-select component doesn't seem to trigger options
  // rendering in tests, so change the input instead
  tree.find("input").simulate("change", { target: { value: "am" } });
  return tree;
};

describe("<AlertManagerInput />", () => {
  it("matches snapshot", () => {
    const tree = ShallowAlertManagerInput();
    expect(tree).toMatchSnapshot();
  });

  it("doesn't render ValidationError after passed validation", () => {
    const tree = ShallowAlertManagerInput();
    silenceFormStore.data.wasValidated = true;
    expect(tree.html()).not.toMatch(/fa-exclamation-circle/);
    expect(tree.html()).not.toMatch(/Required/);
  });

  it("renders ValidationError after failed validation", () => {
    const tree = ShallowAlertManagerInput();
    silenceFormStore.data.alertmanagers = [];
    silenceFormStore.data.wasValidated = true;
    expect(tree.html()).toMatch(/fa-exclamation-circle/);
    expect(tree.html()).toMatch(/Required/);
  });

  it("all available Alertmanager instances are selected by default", () => {
    ShallowAlertManagerInput();
    expect(silenceFormStore.data.alertmanagers).toHaveLength(3);
    for (let i = 1; i <= 3; i++) {
      expect(silenceFormStore.data.alertmanagers).toContainEqual(
        AlertmanagerOption(i)
      );
    }
  });

  it("doesn't override last selected Alertmanager instances on mount", () => {
    silenceFormStore.data.alertmanagers = [AlertmanagerOption(1)];
    ShallowAlertManagerInput();
    expect(silenceFormStore.data.alertmanagers).toHaveLength(1);
    expect(silenceFormStore.data.alertmanagers).toContainEqual(
      AlertmanagerOption(1)
    );
  });

  it("renders all 3 suggestions", () => {
    const tree = ValidateSuggestions();
    const options = tree.find("[role='option']");
    expect(options).toHaveLength(3);
    expect(options.at(0).text()).toBe("am1");
    expect(options.at(1).text()).toBe("am2");
    expect(options.at(2).text()).toBe("am3");
  });

  it("clicking on options appends them to silenceFormStore.data.alertmanagers", () => {
    const tree = ValidateSuggestions();
    const options = tree.find("[role='option']");
    options.at(0).simulate("click");
    options.at(2).simulate("click");
    expect(silenceFormStore.data.alertmanagers).toHaveLength(2);
    expect(silenceFormStore.data.alertmanagers).toContainEqual(
      AlertmanagerOption(1)
    );
    expect(silenceFormStore.data.alertmanagers).toContainEqual(
      AlertmanagerOption(3)
    );
  });

  it("silenceFormStore.data.alertmanagers gets updated from alertStore.data.upstreams.instances on mismatch", () => {
    const tree = ShallowAlertManagerInput();
    alertStore.data.upstreams.instances[0] = {
      name: "am1",
      uri: "http://am1.example.com/new",
      error: ""
    };
    // force update since this is where the mismatch check lives
    tree.instance().componentDidUpdate();
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "am1",
      value: "http://am1.example.com/new"
    });
  });
});
