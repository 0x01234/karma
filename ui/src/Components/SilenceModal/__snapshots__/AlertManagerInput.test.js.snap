// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlertManagerInput /> matches snapshot 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={
    Array [
      Object {
        "label": "am1",
        "value": "http://am1.example.com",
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": 2,
            "name": "Silence form store.alertmanagers[..].keys",
            "observers": Set {},
          },
          "name": "Silence form store.alertmanagers[..]",
          "pendingKeys": Map {
            "cheerio" => false,
            "nodeType" => false,
            "$$typeof" => false,
            "@@__IMMUTABLE_LIST__@@" => false,
            "@@__IMMUTABLE_SET__@@" => false,
            "@@__IMMUTABLE_MAP__@@" => false,
            "@@__IMMUTABLE_STACK__@@" => false,
            "toJSON" => false,
          },
          "proxy": [Circular],
          "target": Object {
            "label": "am1",
            "value": "http://am1.example.com",
            Symbol(mobx administration): [Circular],
          },
          "values": Map {
            "label" => "am1",
            "value" => "http://am1.example.com",
          },
        },
      },
      Object {
        "label": "am2",
        "value": "http://am2.example.com",
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": 2,
            "name": "Silence form store.alertmanagers[..].keys",
            "observers": Set {},
          },
          "name": "Silence form store.alertmanagers[..]",
          "pendingKeys": Map {
            "cheerio" => false,
            "nodeType" => false,
            "$$typeof" => false,
            "@@__IMMUTABLE_LIST__@@" => false,
            "@@__IMMUTABLE_SET__@@" => false,
            "@@__IMMUTABLE_MAP__@@" => false,
            "@@__IMMUTABLE_STACK__@@" => false,
            "toJSON" => false,
          },
          "proxy": [Circular],
          "target": Object {
            "label": "am2",
            "value": "http://am2.example.com",
            Symbol(mobx administration): [Circular],
          },
          "values": Map {
            "label" => "am2",
            "value" => "http://am2.example.com",
          },
        },
      },
      Object {
        "label": "am3",
        "value": "http://am3.example.com",
        Symbol(mobx administration): ObservableObjectAdministration$$1 {
          "defaultEnhancer": [Function],
          "keysAtom": Atom$$1 {
            "diffValue": 0,
            "isBeingObserved": false,
            "isPendingUnobservation": false,
            "lastAccessedBy": 0,
            "lowestObserverState": 2,
            "name": "Silence form store.alertmanagers[..].keys",
            "observers": Set {},
          },
          "name": "Silence form store.alertmanagers[..]",
          "pendingKeys": Map {
            "cheerio" => false,
            "nodeType" => false,
            "$$typeof" => false,
            "@@__IMMUTABLE_LIST__@@" => false,
            "@@__IMMUTABLE_SET__@@" => false,
            "@@__IMMUTABLE_MAP__@@" => false,
            "@@__IMMUTABLE_STACK__@@" => false,
            "toJSON" => false,
          },
          "proxy": [Circular],
          "target": Object {
            "label": "am3",
            "value": "http://am3.example.com",
            Symbol(mobx administration): [Circular],
          },
          "values": Map {
            "label" => "am3",
            "value" => "http://am3.example.com",
          },
        },
      },
    ]
  }
  instanceId="silence-input-alertmanagers"
  isMulti={true}
  onChange={[Function]}
  options={
    Array [
      Object {
        "label": "am1",
        "value": "http://am1.example.com",
      },
      Object {
        "label": "am2",
        "value": "http://am2.example.com",
      },
      Object {
        "label": "am3",
        "value": "http://am3.example.com",
      },
    ]
  }
  placeholder="Alertmanager"
  styles={
    Object {
      "control": [Function],
      "indicatorsContainer": [Function],
      "multiValue": [Function],
      "multiValueLabel": [Function],
      "multiValueRemove": [Function],
      "option": [Function],
      "valueContainer": [Function],
    }
  }
/>
`;
